File: TextShortener.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public class TextShortener 
{
    /// <summary>
    ///     Делим предложение на слова
    /// </summary>
    /// <param name="input"></param>
    /// <returns></returns>
    static List<String> splitTextIntoWords(String input) 
	{
        List<String> words = new ArrayList<>();
        StringBuilder currentWord = new StringBuilder();

        for (char c : input.toCharArray()) 
		{
            if (Character.isLetterOrDigit(c) || c == '\'' || c == '-') {
                currentWord.append(c);
            } 
			else 
			{
                if (currentWord.length() > 0) {
                    words.add(currentWord.toString());
                    currentWord.setLength(0);
                };
            };
        };

        if (currentWord.length() > 0)
            words.add(currentWord.toString());

        return words;
    }

    /// <summary>
    ///     Делим слово по слогам
    /// </summary>
    /// <param name="word"></param>
    /// <returns></returns>
    static List<String> splitWordIntoSyllables(String word) 
	{
        String lowerWord = word.toLowerCase();
        List<Integer> vowelPositions = new ArrayList<>();

        HashSet<Character> vowels = new HashSet<>() {{
            add('а'); add('е'); add('ё'); add('и'); add('о'); add('у'); add('ы'); add('э'); add('ю'); add('я'); // RUS
            add('a'); add('e'); add('i'); add('o'); add('u'); add('y'); // ENG
        }};

        for (int i = 0; i < lowerWord.length(); i++)
            if (vowels.contains(lowerWord.charAt(i))) 
                vowelPositions.add(i);

        if (vowelPositions.isEmpty())
            return new ArrayList<>() {{ add(word); }};

        List<String> result = new ArrayList<>();
        int prevPos = 0;

        for (int i = 0; i < vowelPositions.size(); i++) {
            int currentVowelPos = vowelPositions.get(i);
            if (i > 0) {
                int splitPos = (currentVowelPos + vowelPositions.get(i - 1) + 1) / 2;
                if (word.charAt(splitPos) == 'ь' || word.charAt(splitPos) == 'ъ') splitPos++;
                result.add(word.substring(prevPos, splitPos));
                prevPos = splitPos;
            };
        };
        result.add(word.substring(prevPos));

        return result;
    }

    /// <summary>
    ///     Сокращаем одно слово через дефис
    /// </summary>
    /// <param name="input"></param>
    /// <param name="maxLength"></param>
    /// <returns></returns>
    static String shortenOneWord(String input, int maxLength, String symbol) 
	{
        if (input == null || input.isEmpty() || maxLength <= 0) return "";
        List<String> syllables = splitWordIntoSyllables(input);
        if (syllables.size() < 3) return input.substring(0, maxLength) + "~0";

        maxLength -= symbol.length();

        String fullString = String.join("", syllables);
        int l = -1;
        do {
            if (syllables.size() < 3)
                return input.substring(0, --maxLength) + "~0";

            l = syllables.size() / 2;
            syllables.remove(l);
            fullString = String.join("", syllables);
        } 
		while (fullString.length() > maxLength);

        syllables.add(l, symbol);
        fullString = String.join("", syllables);
        return fullString;
    }

    /// <summary>
    ///     Сокращаем два слова
    /// </summary>
    /// <param name="input"></param>
    /// <param name="maxLength"></param>
    /// <param name="afterwords"></param>
    /// <returns></returns>
    static String shortenTwoWords(String input, int maxLength, int afterwords) 
	{
        maxLength -= (1 + String.valueOf(Math.abs(afterwords)).length());
        if (input.length() < -maxLength) return input + "~" + afterwords;
        String shorten = shortenOneWord(input, maxLength, "-");
        return shorten + "~" + afterwords;
    }
    
    public static String trimString(String input, int maxLength, String space, String symbol) 
	{
        if (input == null || input.isEmpty() || maxLength <= 0) return "";
        List<String> words = splitTextIntoWords(input);
        if (words.isEmpty()) return "";

        if (words.size() == 1) return shortenOneWord(words.get(0), maxLength, "-");
        if (words.size() == 2) return shortenTwoWords(words.get(0), maxLength, 1);

        String fullString = String.join(space, words);
        if (fullString.length() <= maxLength) return fullString;

        int baseCount = words.size();
        int maxWordsLength = maxLength - space.length() * 2 + symbol.length();

        int l = -1;
        do {
            if (words.size() == 2)
                return shortenTwoWords(words.get(0), maxLength, baseCount - 1);

            l = words.size() / 2;
            words.remove(l);
            fullString = String.join(space, words);
        } 
		while (fullString.length() > maxWordsLength);

        words.add(l, symbol);
        fullString = String.join(space, words);
        return fullString;
    }
    
    public static void Test() 
	{
		System.out.println(trimString("Ехал грека через реку, видит грека в реке рак", 30, " ", "..") );
        System.out.println(trimString("Череззаборногузадерищенский кактусообразный параллепипед", 45, " ", "..") );
        System.out.println(trimString("Череззаборногузадерищенский кактусообразный параллепипед", 22, " ", "..") );
        System.out.println(trimString("Череззаборногузадерищенский параллепипед", 22, " ", "..") );
        System.out.println(trimString("Преподавательница", 12, " ", "..") );
	}
	
	public static void main(String[] args) { Test(); }
}

File: TextShortener.js
class TextShortener {
  
    static charIsLetterOrGigit(c)
    {
        return (c.toLowerCase() != c.toUpperCase()) || (!isNaN(c - parseFloat(c)));
    }
  
    /// Делим предложение на слова
  	static splitTextIntoWords(input) 
  	{
		  const words = [];
		  let currentWord = '';
  
		  for (const c of input) 
		  {
  			if (TextShortener.charIsLetterOrGigit(c) || c === '\'' || c === '-') 
			  {
  				currentWord += c;
			  } 
			  else 
			  {
  				if (currentWord.length > 0) {
					  words.push(currentWord);
					  currentWord = '';
				  };
			  };
		  };
  
		  if (currentWord.length > 0) 
		  {
  			words.push(currentWord);
		  };
  
		  return words;
	  }


    ///     Делим слово по слогам
    static splitWordIntoSyllables(word) 
    {
        const lowerWord = word.toLowerCase();
        const vowelPositions = [];

        const vowels = new Set(['а', 'е', 'ё', 'и', 'о', 'у', 'ы', 'э', 'ю', 'я', 'a', 'e', 'i', 'o', 'u', 'y']);

        for (let i = 0; i < lowerWord.length; i++) 
        {
            if (vowels.has(lowerWord[i])) {
                vowelPositions.push(i);
            };
        };

        if (vowelPositions.length === 0) 
        {
            return [word];
        };

        const result = [];
        let previousPosition = 0;

        for (let i = 0; i < vowelPositions.length; i++) 
        {
            const currentVowelPosition = vowelPositions[i];
            if (i > 0) 
            {
                let splitPosition = Math.floor((currentVowelPosition + vowelPositions[i - 1] + 1) / 2);
                if (word[splitPosition] === 'ь' || word[splitPosition] === 'ъ') splitPosition++;
                result.push(word.substring(previousPosition, splitPosition));
                previousPosition = splitPosition;
            };
        };
        result.push(word.substring(previousPosition));

        return result;
    }

    ///     Сокращаем одно слово через дефис
    static shortenOneWord(input, maxLength, symbol = "-") 
    {
        if (!input || maxLength <= 0) return '';
        const syllables = TextShortener.splitWordIntoSyllables(input);
        if (syllables.length < 3) return `${input.substring(0, maxLength)}~0`;

        maxLength -= symbol.length;

        let fullString = syllables.join("");
        let l = -1;
        do 
        {
            if (syllables.length < 3) 
            {
                return `${input.substring(0, --maxLength)}~0`;
            };

            l = Math.floor(syllables.length / 2);
            syllables.splice(l, 1);
            fullString = syllables.join("");
        } 
        while (fullString.length > maxLength);

        syllables.splice(l, 0, symbol);
        fullString = syllables.join("");
        return fullString;
    }

    ///     Сокращаем два слова
    static shortenTwoWords(input, maxLength, afterwords = 1) 
    {
        maxLength -= (1 + Math.abs(afterwords).toString().length);
        if (input.length < -maxLength) return `${input}~${afterwords}`;
        const shorten = TextShortener.shortenOneWord(input, maxLength);
        return `${shorten}~${afterwords}`;
    }

    static trimString(input, maxLength, space = " ", symbol = "..") 
    {
        if (!input || maxLength <= 0) return '';
        const words = TextShortener.splitTextIntoWords(input);
        if (words.length === 0) return '';

        if (words.length === 1) return TextShortener.shortenOneWord(words[0], maxLength);
        if (words.length === 2) return TextShortener.shortenTwoWords(words[0], maxLength);

        let fullString = words.join(space);
        if (fullString.length <= maxLength) return fullString;

        const baseCount = words.length;
        const maxWordsLength = maxLength - space.length * 2 + symbol.length;

        let l = -1;
        do 
        {
            if (words.length === 2) 
            {
                return TextShortener.shortenTwoWords(words[0], maxLength, baseCount - 1);
            };

            l = Math.floor(words.length / 2);
            words.splice(l, 1);
            fullString = words.join(space);
        } 
        while (fullString.length > maxWordsLength);

        words.splice(l, 0, symbol);
        fullString = words.join(space);
        return fullString;
    }

    static test() 
    {
        console.log(TextShortener.trimString("Ехал грека через реку, видит грека в реке рак", 30));
        console.log(TextShortener.trimString("Череззаборногузадерищенский кактусообразный параллепипед", 45));
        console.log(TextShortener.trimString("Череззаборногузадерищенский кактусообразный параллепипед", 22));
        console.log(TextShortener.trimString("Череззаборногузадерищенский параллепипед", 22));
        console.log(TextShortener.trimString("Преподавательница", 12));
    }
}

TextShortener.test();